### PR TITLE
versal2: add initial platform support

### PR DESCRIPTION
**Pull request by Akshay-Belsare**
Add OP-TEE build support for AMD Versal Gen 2 platform. Linux Kernel, optee_os, TF-A, U-Boot are being built outside of build root and sourced from AMD github repo.
Rootfs is from buildroot.

### CODE REVIEW

File: versal2.mk [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+OPTEE_OS_PLATFORM = versal2
+OPTEE_OS_COMMON_EXTRA_FLAGS = CFG_PKCS11_TA=y CFG_USER_TA_TARGET_pkcs11=ta_arm64 O=out/arm
+
+################################################################################
+# Paths to git projects and various binaries
+################################################################################
+TF_A_PATH		?= $(ROOT)/arm-trusted-firmware
+U-BOOT_PATH		?= $(ROOT)/u-boot-xlnx
+LINUX_PATH		?= $(ROOT)/linux-xlnx
+OPTEE_OS_PATH		?= $(ROOT)/optee_os
```
Comment (jenswi-linaro) [inline versal2.mk:34]: You can skip this since it will be assigned the same value `common.mk` below.

File: versal2.mk [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+BR2_PACKAGE_LIBOPENSSL_BIN ?= y
+
+PLATFORM = AMD Versal Gen 2
+OPTEE_OS_PLATFORM = versal2
+OPTEE_OS_COMMON_EXTRA_FLAGS = CFG_PKCS11_TA=y CFG_USER_TA_TARGET_pkcs11=ta_arm64 O=out/arm
+
+################################################################################
+# Paths to git projects and various binaries
+################################################################################
+TF_A_PATH		?= $(ROOT)/arm-trusted-firmware
```
Comment (jenswi-linaro) [inline versal2.mk:31]: Is there any reason why the default values in `common.mk` can't be used?

File: versal2.mk [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+linux-cleaner: linux-cleaner-common
+
+###############################################################################
+# Buildroot
+###############################################################################
+
+BR2_TARGET_GENERIC_ISSUE="OP-TEE embedded distrib for $(PLATFORM)"
+BR2_TARGET_ROOTFS_EXT2=y
+BR2_PACKAGE_BUSYBOX_WATCHDOG=y
```
Comment (jforissier) [inline versal2.mk:155]: `?=` would be preferable

File: versal2.mk [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+all: tfa optee-os dtbo u-boot linux buildroot
+clean: tfa-clean optee-os-clean u-boot-clean linux-clean buildroot-clean dtbo-clean collectartifacts-clean
+
+include toolchain.mk
+
+###############################################################################
+# Collect all artifacts to output directory
+###############################################################################
+
+OUTPUTDIR = ../$(OPTEE_OS_PLATFORM)_output
```
Comment (jenswi-linaro) [inline versal2.mk:51]: A few other platforms call this `BINARIES_PATH`, with the value `$(ROOT)/out` or `$(ROOT)/out/bin`.
Can we use the same convention here?

File: versal2.mk [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+# TF-A, Linux kernel, U-Boot and OP-TEE OS/Client/... are not built from their
+# related Buildroot native package.
+BR2_TARGET_ARM_TRUSTED_FIRMWARE=n
+BR2_LINUX_KERNEL=n
+BR2_TARGET_OPTEE_OS=n
+BR2_TARGET_UBOOT=n
+BR2_PACKAGE_OPTEE_CLIENT=n
+BR2_PACKAGE_OPTEE_TEST=n
+BR2_PACKAGE_OPTEE_EXAMPLES=n
+BR2_PACKAGE_OPTEE_BENCHMARK=n
```
Comment (jforissier) [inline versal2.mk:166]: Not needed I believe (they should in principle be all be disabled by default). Can you test and confirm?
I can see that this copied from stm32mp1.mk, versal.mk and zynqmp.mk.

File: versal2.mk [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+BR2_PACKAGE_LIBP11=y
```
Comment (jforissier) [inline versal2.mk:167]: `?=` and move after line 22 (BR2_PACKAGE_LIBOPENSSL_BIN).

**Review by jforissier — COMMENTED** [review]
Hello @Akshay-Belsare, please see comments below. Thanks!

File: versal2/versal2-memory-reservation.dtso [hunk 7]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: GPL-2.0
```
Comment (jenswi-linaro) [inline versal2/versal2-memory-reservation.dtso:1]: Can we use a permissive license instead, preferably BSD-2-Clause?

File: versal2.mk (hunk 5 above)
Comment (Akshay-Belsare) [reply versal2.mk:166]: @jforissier Tested without these in makefile, working as expected!

File: versal2.mk (hunk 1 above)
Comment (Akshay-Belsare) [reply versal2.mk:34]: @jenswi-linaro
The AMD-Xilinx repositories are arm-trusted-firmware, u-boot-xlnx and linux-xlnx. Will  reuse the OPTEE_OS_PATH from common.mk

File: versal2.mk (hunk 2 above)
Comment (Akshay-Belsare) [reply versal2.mk:31]: We aim to retain and utilize the AMD-Xilinx repository names.
TF-A repo is arm-trusted-firmware
U-boot is u-boot-xlnx
Linux is linux-xlnx

File: versal2.mk [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+################################################################################
+TF_A_PATH		?= $(ROOT)/arm-trusted-firmware
+U-BOOT_PATH		?= $(ROOT)/u-boot-xlnx
+LINUX_PATH		?= $(ROOT)/linux-xlnx
+
+include common.mk
+
+BINARIES_PATH		?= $(ROOT)/out/bin
+TF_A_ELF		?= $(TF_A_PATH)/build/$(OPTEE_OS_PLATFORM)/release/bl31/bl31.elf
+OPTEE_OS_ELF		?= $(OPTEE_OS_PATH)//out/arm/core/tee.elf
```
Comment (jenswi-linaro) [inline versal2.mk:40]: s,//,/

File: versal2.mk [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+	mkdir -p $@
+
+include toolchain.mk
+
+###############################################################################
+# Remove $(BINARIES_PATH) contents and folder
+###############################################################################
+
+binaries-clean:
+	rm -rvf $(BINARIES_PATH)
```
Comment (jenswi-linaro) [inline versal2.mk:64]: This is dangerous, imagine `make BINARIES_PATH=/ binaries-clean`.

File: versal2.mk [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+################################################################################
+# ARM Trusted Firmware
+################################################################################
+
+TF_A_EXPORTS = CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)"
+TF_A_FLAGS = PLAT=versal2 CONSOLE=pl011 RESET_TO_BL31=1 SPD=opteed DEBUG=0 MEM_BASE=0x1600000 MEM_SIZE=0x200000 XILINX_OF_BOARD_DTB_ADDR=0x1000 BL32_MEM_BASE=0x1800000 BL32_MEM_SIZE=0x8000000
```
Comment (jenswi-linaro) [inline versal2.mk:71]: Please break this line, 80 columns is the max.

**Review by jenswi-linaro — COMMENTED** [review]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: versal2.mk [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+################################################################################
+
+LINUX_DEFCONFIG_COMMON_ARCH := arm64
+LINUX_DEFCONFIG_COMMON_FILES := \
+		$(LINUX_PATH)/arch/arm64/configs/xilinx_defconfig \
+		$(CURDIR)/kconfigs/versal2.conf
+
+linux-defconfig: $(LINUX_PATH)/.config
+
+LINUX_COMMON_FLAGS += ARCH=arm64 -j8
```
Comment (etienne-lms) [inline versal2.mk:144]: `-j$(nproc)` would be better. Or maybe let caller decide how many CPU shall participate to the build.

File: versal2.mk [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+	$(U-BOOT_EXPORTS) $(MAKE) -C $(U-BOOT_PATH) clean
+
+###############################################################################
+# Device-Tree
+###############################################################################
+dtbo: u-boot
+	mkdir -p $(BINARIES_PATH)
+	${LINUX_PATH}/scripts/dtc/dtc -@ -I dts \
+		-O dtb -o $(BINARIES_PATH)/versal2-memory-reservation.dtbo \
+		./versal2/versal2-memory-reservation.dtso
```
Comment (etienne-lms) [inline versal2.mk:123]: Suggestion: `$(BUILD_PATH)/versal2/...`

File: versal2.mk [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+linux-cleaner: linux-cleaner-common
+
+################################################################################
+# Buildroot
+################################################################################
+
+BR2_TARGET_GENERIC_ISSUE="OP-TEE embedded distrib for $(PLATFORM)"
+BR2_TARGET_ROOTFS_EXT2=y
+BR2_PACKAGE_BUSYBOX_WATCHDOG=y
```
Comment (etienne-lms) [inline versal2.mk:166]: You missed this comment: s/`...=y`/`... ?= y`/

File: versal2.mk [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+################################################################################
+# U-Boot
+################################################################################
+
+U-BOOT_EXPORTS = CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)"
+U-BOOT_DEFCONFIG_COMMON_FILES := $(U-BOOT_PATH)/configs/amd_versal2_virt_defconfig \
+			$(BUILD_PATH)/kconfigs/u-boot_versal2.conf
+
+
```
Comment (etienne-lms) [inline versal2.mk:99]: Could you remove this extra empty line.

File: versal2.mk [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+LINUX_IMAGE	?= $(LINUX_PATH)/arch/arm64/boot/Image
+ROOTFS_GZ	?= $(ROOT)/out-br/images/rootfs.cpio.gz
+ROOTFS_SIGN	?= $(BINARIES_PATH)/rootfs.cpio.gz.u-boot
+
+################################################################################
+# Targets
+################################################################################
+
+all: tfa optee-os dtbo u-boot linux buildroot buildroot_mkimg
+clean: tfa-clean optee-os-clean u-boot-clean linux-clean buildroot-clean dtbo-clean
```
Comment (etienne-lms) [inline versal2.mk:53]: nitpicking: prefer the same dependency order in `all` and `clean` targets. Not a effective issue, just for consistency.

Comment (Akshay-Belsare) [reply versal2.mk]: Thanks @etienne-lms for the review. I have addressed the comments.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>`

**Pull request merged**